src/application.js

```diff
@@ -0,0 +1,181 @@
+import { CSSResult } from "lit";
+
+import {
+  docTool,
+  domReady,
+  isEmbedded,
+  IS_PRODUCTION,
+  setupLogging,
+  setupHistoryEvents,
+  getMetadataValue,
+} from "./utils";
+import { getReadTheDocsConfig } from "./readthedocs-config";
+import {
+  EVENT_READTHEDOCS_ADDONS_INTERNAL_DATA_READY,
+  EVENT_READTHEDOCS_URL_CHANGED,
+} from "./events";
+
+import * as notification from "./notification";
+import * as analytics from "./analytics";
+import * as search from "./search";
+import * as docdiff from "./docdiff";
+import * as flyout from "./flyout";
+import * as ethicalads from "./ethicalads";
+import * as hotkeys from "./hotkeys";
+import * as linkpreviews from "./linkpreviews";
+import * as filetreediff from "./filetreediff";
+import * as customscript from "./customscript";
+import * as application from "./application";
+import { default as objectPath } from "object-path";
+
+import doctoolsStyleSheet from "./doctools.css";
+
+export class AddonsApplication {
+  constructor() {
+    setupLogging();
+    setupHistoryEvents();
+
+    this.addonsInstances = [];
+    this.config = null;
+
+    console.debug(
+      "Addons Application config (from constructor() method)",
+      this.config,
+    );
+
+    this.addons = [
+      flyout.FlyoutAddon,
+      notification.NotificationAddon,
+      analytics.AnalyticsAddon,
+      ethicalads.EthicalAdsAddon,
+      search.SearchAddon,
+
+      // HotKeys & FileTreeDiff have to be initialized before DocDiff because when
+      // `?readthedocs-diff=true` DocDiff triggers an event that HotKeys has to
+      // listen to to update its internal state.
+      // https://github.com/readthedocs/addons/blob/47645b013724cdf244716b549a5baa28409fafcb/src/docdiff.js#L105-L111
+      hotkeys.HotKeysAddon,
+      filetreediff.FileTreeDiffAddon,
+      docdiff.DocDiffAddon,
+
+      linkpreviews.LinkPreviewsAddon,
+      customscript.CustomScriptAddon,
+    ];
+
+    this.httpStatus = getMetadataValue("readthedocs-http-status");
+
+    this.addDoctoolData();
+    getReadTheDocsConfig(this.sendUrlParam());
+  }
+
+  reload(config) {
+    console.debug("Addons Application config (from reload() method)", config);
+
+    if (!config) {
+      return;
+    }
+    this.config = config;
+    if (this.config && !this.loadWhenEmbedded()) {
+      return;
+    }
+
+    if (!this.addonsInstances.length) {
+      // Addons instances were not created yet
+      let promises = [];
+      for (const addon of this.addons) {
+        if (addon.isEnabled(this.config, this.httpStatus)) {
+          promises.push(
+            new Promise((resolve) => {
+              this.addonsInstances.push(new addon(this.config));
+              resolve();
+            }),
+          );
+        }
+      }
+      Promise.all(promises).catch((err) => {
+        console.error(err);
+      });
```

Comment:
What is the need for promises here? In the previous code we were using `Promise.all().then(...)` and returning this promise from `setup()`. That isn't happening anymore, so promises are probably not needed?

Does this do what you want?

```suggestion
      // Addons instances were not created yet
      try {
        this.addonsInstances = this.addons.filter(
          (addon) => addon.isEnabled(this.config, this.httpStatus);
        ).map(
          (addon) => new addon(this.config);
        );
      } catch err {
        console.error(err);
      });
```

Reply:
> What is the need for promises here?

I thought we wanted to initialize all the addons in parallel -- but it probably doesn't worth the complexity.

> Does this do what you want?

It does. I updated the code to use your proposal.

Reply:
Parallel execution was more important when there were requests taking place before/during instantiation. Without much blocking during instantiation, this is mostly going to randomize the order in which the addons are enabled.

Sequential execution probably helps avoid the issues you noted with load order of the addons.